File: C++17/백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
using namespace std;
int n, cnt;
int col[16];
bool promising(int i) {
	for (int k = 1; k < i; k++) {
		if (col[k] == col[i] || abs(col[i] - col[k]) == abs(i - k))
			return false;
	}
	return true;
}
void Queens(int i) {
	if (promising(i)) {
		if (i == n)
			cnt++;
		else {
			for (int j = 1; j < n + 1; j++) {
				col[i + 1] = j;
				Queens(i + 1);
			}
		}
	}
}
int main() {
	cin >> n;
	cnt = 0;
	Queens(0);
	cout << cnt << "\n";
	return 0;
}

File: PyPy3/백준/Gold/9663. N－Queen/N－Queen.py
def Queens(i):
    if promising(i):
        if i == n:
            global cnt
            cnt += 1
        else:
            for j in range(1, n + 1):
                col[i + 1] = j
                Queens(i + 1)
def promising(i):
    for k in range(1, i):
        if col[k] == col[i] or abs(col[i] - col[k]) == i - k:
            return False
    return True
n = int(input())
cnt = 0
col = [0] * (n + 1)
Queens(0)
print(cnt)